left_sidebar: Show channel folders in left sidebar

evykassirer opened the pull request:
This PR is broken into several commits out of a hope that it makes it easier to review, but let me know if you'd prefer I squash anything (or separate it differently). We should probably squash most of these before merging?

The one thing I still have left to do is add the "N more inactive channels" button for inactive channels. This felt not urgent enough to prevent me from opening this draft PR, but I plan on working on it soon unless we want to punt it a post-release followup (which I think is quite reasonable, considering there will likely be very few inactive channels in folders at first).

I also have barely touched any of the tests and will need to update many of them, but am opening this PR for us to test-deploy on CZO in the meantime.

Here are some test cases I've tested manually. It would be great for a reviewer to also run through most of these. Feel free to also add further test cases to this list!

- pinned channels section
    - muted channels are still faded
    - pinned inactive channels show up (notably without the view N more logic) and aren’t faded
    - pinned channels that are also in channel folders are displayed in the pinned section
    - pinning and unpinning a channel updates the list immediately (including adding/removing the whole pinned channels section if it transitions to or from being empty)
- channel folder sections section
    - muted channels still faded
    - inactive channels in the folder show up (after clicking view N more) and aren’t faded
        - **TODO**: add the "view N more" functionality
    - adding and removing a channel to a folder updates the stream list immediately (including adding/removing a channel folder section if it transitions to or from being empty)
- "other channels” section
    - should be named just “channels” if there are no folders or pinned channels
- inactive section
    - all channels are faded (no visual distinction for muted channels)
- plus button for channel folders
    - only visible for channel folders and the “other channels” section (not pinned or inactive)
    - doesn’t show up for users who can’t create channels (e.g. guests, spectators)
    - tooltip says “Create a channel” not “Add channels”
    - clicking the button opens the creation modal with the folder selected (or no folder selected, for the “other channels” section)
- search filter
    - X button only visible when text is in the search box
    - filtering works as expected (only matches are visible)
    - clearing search with X or ESC works
- plus button beside search filter
    - not visible when logged out (and therefore filter becomes full width)
    - tooltip says “Create a channel” not “Add channels”
    - clicking it opens the popover if two options are available, and just opens the browse channels overlay if channel creation isn’t permitted
- sticky on scroll
    - search bar is always visible on scroll, and nothing pokes out behind it as it scrolls past
    - other headers stick until the section has scrolled past and another header replaces it
- unread counts
    - section counts are correct totals of their channels’ counts
    - counts are hidden when “Show unread count summaries in the left sidebar” isn’t checked, and show again on hover
    - when the unread count is 0, it's not shown (and dots aren't shown when counts are hidden)
- misc
    - sections can be toggled collapsed/uncollapsed by clicking on them
    - when pinning a channel, if the pinned section isn’t in view, the stream list scrolls so that the pinned channel comes into view
    - when putting a channel in a folder, if the folder isn’t in view, the stream list scrolls so that the moved channel comes into view

alya commented:
The "+" button on channel folders is misaligned when there are no unreads.

alya commented:
Folders aren't collapsing when I click on them.

alya commented:
Can we make inactive channels folded by default?

alya commented:
When I created an inactive channel, for some reason it took quite a while for it to start looking faded in the sidebar.

evykassirer commented:
> Folders aren't collapsing when I click on them.

They were working sometimes, but I realized this stopped when we redraw the stream list. Fixed!

Thanks for the other finds, I'll work on those shortly.

alya commented:
Yep, just realized that and fixed my comment. :)

evykassirer commented:
> The "+" button on channel folders is misaligned when there are no unreads.

I was aiming to have it aligned with the unread counts, the same way we do with topics in a channel. Do you want something different than that?

alya commented:
The sidebar scrolls when I collapse a folder. E.g. before collapsing:

![Screenshot 2025-07-01 at 11 47 33](https://github.com/user-attachments/assets/28bc2b3d-2c18-4bb1-bf66-5819134f8f3a)

After:

![Screenshot 2025-07-01 at 11 47 37](https://github.com/user-attachments/assets/76bfd3d3-1e5f-4a19-b5c5-814dc54f2a49)

The moves can be more dramatic, too, if the folder has lots of content.

I think the folder name should remain in place, and just the stuff below should scroll up to it.

evykassirer commented:
I think the unread markers issue(s) should be fixed now!

timabbott commented on web/src/stream_events.ts:
```diff
@@ -146,6 +146,9 @@ export function update_property<P extends keyof UpdatableStreamProperties>(
             recent_view_ui.complete_rerender();
+            // Since muted streams aren't counted in visible unread
+            // counts, we need to update the rendering of them.
+            unread_ui.update_unread_counts();
```
I think it might make more sense to either have all of this code be in `stream_muting.update_is_muted` or no UI update code in that other module, i.e. inlining it here and deleting that one-function file.

The main thing that's messy is that `message_lists.all_rendered_message_lists` loop, and that surely could be made a function in `message_live_update.ts`.

Would be a good little prep PR.

evykassirer replied:
If that worked, I can also make that a prep PR

evykassirer replied:
Yup it worked! #35107

evykassirer replied:
Okay for the sake of time I'm going to remove this commit and leave it as it is, and add `unread_ui.update_unread_counts();` to `build_stream_list` for this PR since that ensures the stream list gets built correctly. I can look into this wider refactor in the future.

timabbott commented on web/src/hotkey.js:
```diff
@@ -588,14 +588,6 @@ export function process_enter_key(e) {
-            // to close stream search.
-            stream_list.clear_and_hide_search();
-            return true;
-        }
```
Hmm, so this is about `Esc` when you have something typed in stream search. Are you sure we don't want that to still clear it, even though it can't be hidden now? I feel like we might have other inputs that would be useful guidance.

timabbott commented on web/src/stream_list.ts:
```diff
@@ -1044,33 +1047,18 @@ export function searching(): boolean {
 }
 
-export function clear_search(e: JQuery.ClickEvent): void {
-    e.stopPropagation();
```
Do you know what this `stopPropagation` was here? It's not obvious to me why it should or should not still be.

evykassirer replied:
`clear_search` was being called for a click, I moved the `stopPropagation` to `$("#clear_search_stream_button").on("click", (e: JQuery.ClickEvent) => {` below

evykassirer commented:
> Add topic "+" button tooltip is getting layered wrong for inactive streams

Fixed in this most recent push

evykassirer commented:
Update: If you collapse a channel folder while you're in a channel in that folder, keep showing the folder/topics underneath the folder heading, just like we do for the DM conversation you're viewing when the DMs section is collapsed.

![Kapture 2025-07-02 at 14 40 34](https://github.com/user-attachments/assets/d5adf145-c8de-4d37-a5f3-82b8d03dc6fd)

evykassirer commented:
Updates: two new commits from [#design > channel folders: unread counters @ 💬](https://chat.zulip.org/#narrow/channel/101-design/topic/channel.20folders.3A.20unread.20counters/near/2211570)

* left_sidebar: Only show channel section unread count when collapsed.
* left_sidebar: Show faded muted unread for header if no regular unreads.

timabbott commented on web/src/stream_list_sort.ts:
```diff
@@ -71,13 +68,16 @@ export function has_recent_activity(sub: StreamSubscription): boolean {
-    dormant_streams: number[];
-    muted_pinned_streams: number[];
-    muted_active_streams: number[];
+    sections: StreamListSection[];
```
The refactoring in this file is pretty nice!

timabbott commented on web/src/stream_list.ts:
```diff
@@ -268,88 +268,67 @@ export function build_stream_list(force_rerender: boolean): void {
+            add_sidebar_li(stream_id, $(`#stream-list-${section.id}`));
+        }
     }
+    unread_ui.update_unread_counts();
```
This is a big hammer that can be quite expensive; do we really need to call it here? It recomputes and rerenders unread counts everywhere in the UI, which mean every time we build this component we'll trigger rebuilding a bunch of other components too.

I would have thought this component would already be taking care of presenting accurate initial unread counts when built; I at least don't see a similar call pre-patch.

evykassirer replied:
Okay I figured it out -- we do need to do it because when we rerender the sidebar, we need to decide which counts to show and which to hide while rendering, which to mark as a muted count, etc. Either we should store all that data (which would be more complex) or we can update the counts again. I'm adding it back.

timabbott replied:
I think there's likely a much smaller thing you can call. The right sidebar does not need a rerender, for example. 

The computation is  `unread.get_counts`, if that's what you need. But we surely already call that for rendering the left sidebar. So I think the issue is you just don't have the data flow right to use the existing call.

timabbott replied:
OK. We should file an issue for 12.0 to look at optimizing how much of the view-change time can be optimized; I feel like the sidebars are now like 40% of the time and probably should not be.

timabbott commented on web/templates/left_sidebar.hbs:
```diff
@@ -186,11 +186,16 @@
-                <button type="button" class="clear_search_button" id="clear_search_stream_button">
-                    <i class="zulip-icon zulip-icon-close" aria-hidden="true"></i>
-                </button>
+            <div id="stream_search_and_add" class="zoom-in-hide">
```
Remember to use dashes, not underscores, whenever you're picking a new variable name.

evykassirer replied:
I've updated a few and will try to remember this next time! Thanks

evykassirer commented on web/src/stream_list.ts:
```diff
@@ -283,14 +283,21 @@ export function build_stream_list(force_rerender: boolean): void {
+            can_create_streams &&
+            !["pinned-streams", "dormant-streams"].includes(section.id)
+        ) {
+            plus_icon_url = `#channels/folders/${section.id}/new`;
```
I don't currently have anything on the `section` that makes checking if it's a folder straightforward (the id is just a number instead of the strings).

Unless you'd prefer I add `is_folder_section` or something to the section data, I'll keep it as the inverse check.

```js
function get_section_channel_plus_icon_url(section: StreamListSection): string | undefined {
    if (section.id === "normal-streams") {
        return "#channels/new";
    } else if (
        !["pinned-streams", "dormant-streams"].includes(section.id)
    ) {
        return `#channels/folders/${section.id}/new`;
    }
    return undefined;
}

// ...

    plus_icon_url: can_create_streams ? get_section_channel_plus_icon_url(section) : undefined,

```

timabbott replied:
In #35152, we're using negative fake folder IDs for those sections internally ... but I wouldn't want those in the DOM.

evykassirer replied:
Resolving, feel free to reopen if you still have concerns :)

evykassirer commented:
Updates:
* Fix [#design > channel folders: show all topics UI @ ](https://chat.zulip.org/#narrow/channel/101-design/topic/channel.20folders.3A.20show.20all.20topics.20UI/near/2214634)💬
* show inactive channels in folder sections on search (without any visual "show/hide" button)

Still to do:
* [ ] show a summary counter on the "show inactive" button too, like we do for topics
* [ ] When using keyboard navigation to move down the list of channels, skip over inactive channels when they're hidden -- move keyboard focus to the show/hide button and then to the next section.
* [ ] if a section is collapsed, skip over those elements for keyboard navigation

karlstolley commented on web/styles/left_sidebar.css:
```diff
@@ -313,6 +313,73 @@
+    position: sticky;
+    top: 0;
+    /* Must be more than .stream-list-subsection-header */
+    z-index: 3;
```
This is a follow-up for after this PR gets merged, but the use of `z-index` in situations like this is regrettable. What we need, I think, is to better organize elements in the DOM and use `position: sticky` with thoughtful `top:` values that allow things to stack up the way we want them to.

evykassirer replied:
Yeah this is mostly piggybacking on what was already there, but agreed that it would be better to not use `z-index`.

karlstolley commented on web/styles/left_sidebar.css:
```diff
@@ -313,6 +313,73 @@
+    position: sticky;
+    /* Input is 29px tall at 16px/1em; plus 2px for top/bottom border.
+       Input container has 8px and 3px top/bottom padding = 11px total padding */
+    top: calc(1.8125em + 2px + 11px);
```
I corrected this value this afternoon; left a note on the commit for @evykassirer's sake, but that can be removed once they've seen it.

evykassirer replied:
Thanks!

karlstolley commented on web/styles/left_sidebar.css:
```diff
@@ -345,11 +345,17 @@
             --line-height-sidebar-row-prominent
         )
-        / var(--right-sidebar-header-icon-toggle-width) minmax(0, 1fr);
+        / var(--right-sidebar-header-icon-toggle-width) minmax(0, 1fr) minmax(
```
Actually, I stand corrected. These are the same. It's the marker that's making it look off.

evykassirer replied:
Hm, not reserving a specific amount of space for the unread count also puts us in situations like this

<img width="328" height="97" alt="image" src="https://github.com/user-attachments/assets/6c3fe88c-1608-4319-87db-2d257fa9fa44" />

evykassirer replied:
@karlstolley let me know if you still have concerns, resolving for now

karlstolley commented on web/styles/left_sidebar.css:
```diff
@@ -372,6 +390,10 @@
         }
+
+        .add_stream_icon {
+            display: grid;
```
This isn't accurate...there's no grid defined here. Would it be okay to just use `visibility` on `.add_stream_icon`? I'd think so.

evykassirer replied:
Yup that works, changed to that thanks :)

evykassirer commented on web/styles/left_sidebar.css:
```diff
@@ -360,6 +366,18 @@
+        text-decoration: none;
+        grid-auto-columns: var(--left-sidebar-header-icon-width);
+        grid-template-rows: calc(
+            var(--line-height-sidebar-row-prominent) - 4px
```
Subtracting the 2px of margin on each side, I'll add a comment.

evykassirer commented:
Updates:

* Show count on hover on folder headings
* Fixed [#issues > 🎯 channel folder toggle not responding after zoom in @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.8E.AF.20channel.20folder.20toggle.20not.20responding.20after.20zoom.20in/near/2221580)
* Removed inactive section (put that in its own commit for now)
*

evykassirer commented:
Updates:

* Show count on hover on folder headings
* Fixed [#issues > 🎯 channel folder toggle not responding after zoom in @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.8E.AF.20channel.20folder.20toggle.20not.20responding.20after.20zoom.20in/near/2221580)
* Removed inactive section (put that in its own commit for now)
* Removed 10px of margin between headings

evykassirer commented:
Update:

* `fixup: reduce what's shown when in a channel within a collapsed folder`

evykassirer commented:
updates:

* Fixed some unread bugs (the inactive button's unread count wasn't displaying in grid, we weren't showing muted counts sometimes when we should have)
* Addressed [this todo](https://github.com/zulip/zulip/pull/35075#discussion_r2193621538), which should ideally be integrated into other commits but I pushed the refactor as its own commit for now.

evykassirer commented:
Updates:

* Added ` left_sidebar: Skip collapsed sections during keyboard navigation of streams` ([diff](https://github.com/zulip/zulipcompare/27b3494e768b87ac4a64f38084fd151f39763591..97413c2d158d2ec64402fc507b9a1b44c6058b72)) (no selecting of inactive toggle yet)
* Rebased and resolved conflicts

evykassirer commented on web/src/stream_list_sort.ts:
```diff
@@ -218,38 +261,70 @@ export function sort_groups(stream_ids: number[], search_term: string): StreamLi
+            const stream_id = prev_row.stream_id;
+            assert(stream_id !== undefined);
+            const section = section_id_map.get(stream_id)!;
+            if (collapsed_sections.has(section.id)) {
```
So... this is currently skipping channels that we keep shown because they're active when a section is collapsed. Not sure what to do about that yet.

amanagr replied:
Added a commit for that.

evykassirer commented:
Updates:

* Fixed [unread count update implementation](https://github.com/zulip/zulip/pull/35075#discussion_r2229358878).
* Channel folders now don't uncollapse in any situation other than clicking the folder or clicking an active channel visible in the collapsed folder

evykassirer commented:
As far as I know, this PR now has 0 known bugs or issues to work (other than fixing unit tests) 🎉

timabbott commented on web/src/stream_list_sort.ts:
```diff
@@ -23,6 +23,25 @@ export function get_stream_ids(): number[] {
+    return map;
+}
+
+export function current_section_id_for_stream(stream_id: number): string {
```
This function is a bit dangerous for performance; it sounds like it should be `O(1)`, but is actually `O(N)`. 

Probably worth a comment: 

> Warning: This function is O(n), so it should not be called in a loop.

I checked all the existing callers are fine on that concern.

timabbott commented on web/src/channel_folders.ts:
```diff
@@ -37,6 +37,10 @@ export function get_channel_folders(include_archived = false): ChannelFolder[] {
 
+export function get_channel_ids(): number[] {
+    return [...channel_folder_by_id_dict.keys()];
+}
```
Hmm, what are the semantics of this function -- is this the IDs of just those channels that are in a folder?

Probably a `get_channel_ids_with_folders` name would clarify that.

timabbott replied:
Wait, how you're using it seems to expect this is a set of folder_ids, this seems more confused...

evykassirer replied:
Yup it should be folder ids, my mistake. Thanks for noticing

timabbott replied:
``` diff
diff --git a/web/src/channel_folders.ts b/web/src/channel_folders.ts
index 789eb5b532..7d6b66e69f 100644
--- a/web/src/channel_folders.ts
+++ b/web/src/channel_folders.ts
@@ -37,7 +37,7 @@ export function get_channel_folders(include_archived = false): ChannelFolder[] {
     });
 }
 
-export function get_channel_ids(): number[] {
+export function get_all_folder_ids(): number[] {
     return [...channel_folder_by_id_dict.keys()];
 }
 
diff --git a/web/src/left_sidebar_navigation_area.ts b/web/src/left_sidebar_navigation_area.ts
index 0c4bdaa4c4..075da59d8e 100644
--- a/web/src/left_sidebar_navigation_area.ts
+++ b/web/src/left_sidebar_navigation_area.ts
@@ -179,7 +179,7 @@ export function update_dom_with_unread_counts(
         inactive_section_unread_counts.muted,
     );
 
-    for (const folder_id of channel_folders.get_channel_ids()) {
+    for (const folder_id of channel_folders.get_all_folder_ids()) {
         const unread_counts = folder_unread_counts.get(folder_id);
         update_section_unread_count(
             $(`#stream-list-${folder_id}-container .stream-list-subsection-header`),
```

timabbott replied:
It's be a nice tweak to make this a `Set` instead.

timabbott commented on web/src/stream_list.ts:
```diff
@@ -281,6 +280,15 @@ export function rewire_stream_list_section_container_html(
+    } else if (!["pinned-streams", "dormant-streams"].includes(section.id)) {
+        return `#channels/folders/${section.id}/new`;
+    }
+    return undefined;
```
I feel like this would probably be nicer if `section` just had a `folder_id` field, that's null for non-folder sections, and then just wrote 
```
if (section.folder_id !== null) { 
    return `#channels/folders/${section.folder_id}/new`;
} else if (section.id="normal-streams") { 
    return "#channels/new"; 
}
return undefined;
```

So much more readable locally without having to know what the complete set of non-channel-folders are.

timabbott commented on web/src/stream_list_sort.ts:
```diff
@@ -156,8 +176,21 @@ export function sort_groups(stream_ids: number[], search_term: string): StreamLi
+        pinned_section.muted_streams.length === 0
+    ) {
+        normal_section.section_title = $t({defaultMessage: "CHANNELS"});
+    }
```
Does this mean if we only have "muted channels" and "non-muted channels" that their names are "OTHER CHANNELS" and "MUTED CHANNELS"? That may deserve attention in #design.

evykassirer replied:
"MUTED CHANNELS" isn't a typical folder name

timabbott commented on web/src/stream_settings_ui.ts:
```diff
@@ -243,6 +244,13 @@ export function update_is_default_stream(): void {
+    const $stream_li = stream_list.get_stream_li(sub.stream_id);
+    if ($stream_li) {
+        stream_list.scroll_stream_into_view($stream_li);
+    }
```
Why are we scrolling the channel into view from here? This seems very odd, since it's an administrator action to edit channel folder membership.

evykassirer replied:
Oh, this is leftover from my misunderstanding that they were personal folders. Thanks for catching this!

timabbott commented on web/src/left_sidebar_navigation_area.ts:
```diff
@@ -167,6 +179,15 @@ export function update_dom_with_unread_counts(
+        update_section_unread_count(
+            $(`#stream-list-${folder_id}-container .stream-list-subsection-header`),
+            unread_counts?.unmuted ?? 0,
+            unread_counts?.muted ?? 0,
```
I think the better answer would be to replace `const unread_counts = folder_unread_counts.get(folder_id);` with something that returns a a properly typed empty folder object. Or just having a loop at the beginning that creates those.

timabbott commented on web/src/stream_list.ts:
```diff
@@ -322,17 +332,34 @@ export function build_stream_list(force_rerender: boolean): void {
+            );
+        }
     }
     // Rerendreing can moving channels between folders and change heading unread counts.
```
Type, should be `Rerendering`

timabbott commented on web/styles/left_sidebar.css:
```diff
@@ -403,6 +405,67 @@
+
+    .show-inactive-channels,
+    .hide-inactive-channels {
+        /* Override the action heading font size, so that em measurements
```
This is for a follow-up, but I wonder whether the "action heading" font size should be inheriting to avoid this problem. I'm not entirely sure which font size you're talking about here.

timabbott commented on web/src/stream_list_sort.ts:
```diff
@@ -225,36 +254,72 @@ export function sort_groups(stream_ids: number[], search_term: string): StreamLi
     }
-
-    return maybe_get_stream_id(i + 1);
+    return undefined;
```
The core logic of this could be extracted as a boolean function `is_visible_row` and shared between this function and `prev_row_id` -- you just return it if `true`, `continue` if false.

timabbott replied:
``` diff
diff --git a/web/src/stream_list.ts b/web/src/stream_list.ts
index 29ad358946..be68b5852c 100644
--- a/web/src/stream_list.ts
+++ b/web/src/stream_list.ts
@@ -884,7 +884,7 @@ export function initialize_stream_cursor(): void {
                     topic_list.active_stream_id(),
                 ),
             next_key: (row) =>
-                stream_list_sort.next_row_id(
+                stream_list_sort.next_row(
                     row,
                     sections_showing_inactive,
                     collapsed_sections,
diff --git a/web/src/stream_list_sort.ts b/web/src/stream_list_sort.ts
index 81be5ca497..30139f01b0 100644
--- a/web/src/stream_list_sort.ts
+++ b/web/src/stream_list_sort.ts
@@ -259,6 +259,32 @@ export function first_row(): StreamListRow | undefined {
     return all_rows.at(0);
 }
 
+function is_visible_row(
+    row: StreamListRow,
+    section_id_map: Map<number, StreamListSection>,
+    sections_showing_inactive: Set<string>,
+    collapsed_sections: Set<string>,
+    active_stream_id: number | undefined) {
+    if (row.type === "stream") {
+        const stream_id = row.stream_id;
+        assert(stream_id !== undefined);
+        const section = section_id_map.get(stream_id)!;
+        if (collapsed_sections.has(section.id) && active_stream_id !== stream_id) {
+            return false;
+        }
+        if (!sections_showing_inactive.has(section.id) && row.inactive) {
+            return false;
+        }
+    } else if (
+        row.type === "inactive_toggle" &&
+            collapsed_sections.has(row.section_id)
+    ) {
+        return false;
+    }
+    return true;
+
+}
+
 export function prev_row(
     row: StreamListRow,
     sections_showing_inactive: Set<string>,
@@ -270,28 +296,14 @@ export function prev_row(
     while (i > 0) {
         i -= 1;
         const prev_row = all_rows[i]!;
-        if (prev_row.type === "stream") {
-            const stream_id = prev_row.stream_id;
-            assert(stream_id !== undefined);
-            const section = section_id_map.get(stream_id)!;
-            if (collapsed_sections.has(section.id) && active_stream_id !== stream_id) {
-                continue;
-            }
-            if (!sections_showing_inactive.has(section.id) && prev_row.inactive) {
-                continue;
-            }
-        } else if (
-            prev_row.type === "inactive_toggle" &&
-            collapsed_sections.has(prev_row.section_id)
-        ) {
-            continue;
+        if (is_visible_row(prev_row, section_id_map, sections_showing_inactive, collapsed_sections, active_stream_id)) {
+            return prev_row;
         }
-        return prev_row;
     }
     return undefined;
 }
 
-export function next_row_id(
+export function next_row(
     row: StreamListRow,
     sections_showing_inactive: Set<string>,
     collapsed_sections: Set<string>,
@@ -302,23 +314,9 @@ export function next_row_id(
     while (i + 1 < all_rows.length) {
         i += 1;
         const next_row = all_rows[i]!;
-        if (next_row.type === "stream") {
-            const stream_id = next_row.stream_id;
-            assert(stream_id !== undefined);
-            const section = section_id_map.get(stream_id)!;
-            if (collapsed_sections.has(section.id) && active_stream_id !== stream_id) {
-                continue;
-            }
-            if (!sections_showing_inactive.has(section.id) && next_row.inactive) {
-                continue;
-            }
-        } else if (
-            next_row.type === "inactive_toggle" &&
-            collapsed_sections.has(next_row.section_id)
-        ) {
-            continue;
+        if (is_visible_row(next_row, section_id_map, sections_showing_inactive, collapsed_sections, active_stream_id)) {
+            return next_row;
         }
-        return next_row;
     }
     return undefined;
 }
```

timabbott commented on web/src/stream_list.ts:
```diff
@@ -863,19 +863,35 @@ const update_streams_for_search = _.throttle(actually_update_streams_for_search,
+                }
+                return $(
+                    `#stream-list-${opts.key.section_id}-container .stream-list-toggle-inactive-channels`,
+                );
```
This would be better written with a switch on `key.type` with a `default: throw new Error`. 

The main flaw is that having an unknown type result in our assuming it is an inactive channels elements is bad. A few reasons:
- It will silently break if a new type is added.
- Searching/grepping for the inactive-channel type won't find this code.
- The code is less readable, since it's implicitly asserting that a `opts.key` is a `StreamRow`. As a side note, why does `opts` not have a type annotation?

evykassirer replied:
`opts` doesn't need a type annotation, it's implied by `ListCursor`

timabbott commented on web/src/stream_list.ts:
```diff
@@ -981,6 +981,11 @@ function on_sidebar_channel_click(
         e.stopPropagation();
     }
 
+    const section_for_stream = stream_list_sort.current_section_id_for_stream(stream_id);
```
This probably deserves a comment of the form:

> In the event that user clicks on the channel when in a collapsed folder, which is only possible if the user was already viewing that channel, we uncollapse the section because ...

alya commented:
I don't understand why the counter says 3 here. Seems like it should be 2.

<img width="312" height="241" alt="Screenshot 2025-07-25 at 17 43 58" src="https://github.com/user-attachments/assets/f64524a3-0941-4608-a539-8bd8ae8b66b8" />

timabbott commented:
I resolve some of my notes above, and also fixed about half of the tests. Remaining TODOs before we can merge:
- `stream_list_sort.ts` tests fail starting with "left_sidebar: Show channel folders in left sidebar"; I changed some of them but got interrupted before finishing.
- `strema_list` tests fail earlier.
- Some tests disabled with `git grep TODO/channel` comments.
- The various unresolved comments above; a couple of which are about functional concerns.

amanagr commented:
Fixed the CI

evykassirer commented on web/styles/left_sidebar.css:
```diff
@@ -88,7 +88,8 @@
 
-#left-sidebar-navigation-list .selected-home-view {
+#left-sidebar-navigation-list .selected-home-view,
+.show-inactive-channels {
```
Unread messages are shown (and sometimes masked) on the show inactive channels button, in very similar ways to the headers and the home view. See `update_section_unread_count` for how it's toggled.

I think there's something funky going on here with `hide_unread_counts` vs `hide-unread-messages-count`. I don't have a clear sense of which is used when and why, and I think the inactive channel CSS should be in the same place as the header CSS ideally. So this is worth looking more into, though I also think there aren't currently any visual bugs around (as far as I know).

evykassirer commented on web/tests/stream_list_sort.test.cjs:
```diff
@@ -193,13 +176,7 @@ test("basics", () => {
-        assert.deepEqual(sorted_sections.length, 3);
-=======
-        assert.deepEqual(sorted_sections.length, 2);
->>>>>>> e3f8b5fa9f (left_sidebar: Remove inactive section, put inactive channels in regular section.)
```
Not ideal to have this in the commits, but probably not worth going back to fix it?
